### PR TITLE
Allow sending of any combination of text and html

### PR DESCRIPTION
**Pull request by TimQuelch**
- Add a new property to messages: `format` which can take a value of
'text', 'html', or 'both'. 'html' is similar to the old default
behaviour, 'both' is similar to the old functionality when
`org-msg-text-plain-alternative` was non-nil, and 'text' sends only
text/plain and no 'html'
- Add a custom option to configure the default value of this property
- Refactor `org-msg-prepare-to-send` to handle the new functionality
- Also refactors `org-msg-mml-into-multipart-related` so that MIME
structures are as close as possible to when recursive mime is present.

Creating this PR now that it is in a mostly working state

My knowledge of mml and MIME are not very deep, so please let me know if I have made any mistakes with the refactoring of `org-msg-prepare-to-send` or with `org-msg-mml-into-multipart-related`.

This patch should work with #69 to allow for much greater functionality and configurability with sending text email while still remaining in `org-msg`

TODO
- Add command to cycle through different `format` properties for easy switching between text, html, and both
- ~Test when `mml-expand-all-html-into-multipart-related` is present~ DONE, seems to work

### CODE REVIEW

File: org-msg.el [hunk 1]
```diff
@@ -731,27 +750,59 @@ This function is a hook for `message-send-hook'."
...
-	  (mml-insert-part "text/html")
-	  (insert (propertize (org-msg-xml-to-str mail) 'mml t)))))))
+
+          ;; If mml has recursive html support (in later versions of emacs)
+          ;; we want to generate the structure of the MIME document here.
+          ;; If not we do this by manually editing the structure of the
+          ;; parsed MML tree in `org-msg-mml-into-multipart-related'. We
+          ;; also don't need to worry about this if we are only sending
+          ;; text/plain
+          (if (and (not (org-msg-mml-recursive-support))
```
Comment (jeremy-compostella) [inline org-msg.el:774]: ¬A and ¬B ⇔ ¬(A or B)
Then if you reverse the logic you end with a simple 

```
(if (or (org-msg-mml-recursive-support)
        (eq org-msg-current-message-format 'text)))
``` 

which is easier to understand => easier to maintain.

File: org-msg.el [hunk 2]
```diff
@@ -155,8 +158,14 @@ It is used by function advice.")
   "Org Mode #+STARTUP."
   :type '(string))
 
-(defcustom org-msg-text-plain-alternative nil
-  "Include an ASCII export as a text/plain alternative.")
+(defcustom org-msg-default-message-format 'html
```
Comment (jeremy-compostella) [inline org-msg.el:161]: This should be a list which would make `both` useless and also allow to include more exotic alternatives if need be. A element of the list could be tested with `assoc` and it can be iterated on. Also, a list has an order allowing the user to customize the order of the alternatives.

```
(defcustom org-msg-alternatives '(html)
[...]
```

Then I would recommend using the template method pattern like below (and this just me thinking out loud so it certainly needs some improvement and to point to actual functions):
```
(defconst org-msg-alernative-to-mime
          '((text (text/plain . org-msg-export-to-plain-text)
            (html (text/html . org-msg-export-to-html))
          "Associative list of alternatives")
``` 
The documentation of `org-msg-alternatives` could then point to `org-msg-alernative-to-mime` for the list of supported alternatives. Note that we could use a simple list of association but we would have to use the MIME type directly in org-msg-alternatives which could be problematic later if someone wants to use various types for different text exportation like for instance include org-mode text or ascii-utf16 ...

File: org-msg.el [hunk 3]
```diff
@@ -731,27 +750,59 @@ This function is a hook for `message-send-hook'."
...
+          (setq org-msg-current-message-format (org-msg-get-message-format))
+          ;; Ensure that all attachments exist
 	  (dolist (file attachments)
 	    (unless (file-exists-p file)
 	      (error "File '%s' does not exist" file)))
+          ;; Set temporary global variable with attachment list
 	  (setq org-msg-attachment attachments)
-	  (when org-msg-text-plain-alternative
+          ;; Generate the text version if required
+	  (when (not (eq org-msg-current-message-format 'html))
```
Comment (jeremy-compostella) [inline org-msg.el:762]: `(when (not [...]`  ⇔ `(unless [...])`

File: org-msg.el [hunk 4]
```diff
@@ -731,27 +750,59 @@ This function is a hook for `message-send-hook'."
...
 	    (setq org-msg-text-plain (org-msg-org-to-text-plain)))
+          ;; Clear the contents of the message
```
Comment (jeremy-compostella) [inline org-msg.el:764]: Indentation.

File: org-msg.el [hunk 5]
```diff
@@ -770,22 +821,28 @@ This function is used as an advice function of
...
   (setq cont (funcall orig-fun cont))
-  (let ((newparts '()))
+  (let ((attachmentparts '()))
+    ;; Generate this list of attachment parts
     (dolist (file org-msg-attachment)
       (let ((type (org-msg-file-mime-type file)))
 	(push (list 'part `(type . ,type) `(filename . ,file)
 		    '(disposition . "attachment"))
-	      newparts)))
+	      attachmentparts)))
```
Comment (jeremy-compostella) [inline org-msg.el:830]: If you believe a renaming is necessary submit a separate patch it has nothing to do with what you are trying to achieve here.

File: org-msg.el [hunk 6]
```diff
@@ -731,27 +750,59 @@ This function is a hook for `message-send-hook'."
...
+                (forward-line))         ; Exit text/html
+              (when (eq org-msg-current-message-format 'both)
+                (forward-line))         ; Exit alternative
+              (when attachments
+                (dolist (file attachments)
+		  (mml-insert-tag 'part
+                                  'type (org-msg-file-mime-type file)
+				  'filename file
+                                  'disposition "attachment")))))
+
```
Comment (jeremy-compostella) [inline org-msg.el:800]: I find this area overly complicated.  I believe that switching to a list should help make this code look better.

**Review by jeremy-compostella — COMMENTED** [review]
Thanks for the submission I really appreciate it. This is a nice idea.
I have some implementation suggestion that I hope you will like too.

A few general notes:
1. I would prefer that you squash patches fixing this patch into a unique patch. It has not been merged in an official branch yet so we do not need to remember that you originally had a typo.
2. If you want to make patches which are unrelated to this feature (like renaming variable you don't really touch or adding comments) please do so in different patches but feel free to all your patches in a unique  pull request.
3. Include you sign-off in the commit message
4. indent the lines in your commit message

File: org-msg.el (hunk 2 above)
Comment (TimQuelch) [reply org-msg.el:161]: I like this idea. I'll have a look at implementing something like this

File: org-msg.el (hunk 5 above)
Comment (TimQuelch) [reply org-msg.el:830]: This rename was something I did just as part of my process of understanding what the function was actually doing. I can revert the change though

File: org-msg.el (hunk 6 above)
Comment (TimQuelch) [reply org-msg.el:800]: Agreed. I couldn't think of an easy way to simplify it while still behaving correctly. We'll see how it looks with the list config

**Comment by TimQuelch** [comment]
> I would prefer that you squash patches fixing this patch into a unique patch. It has not been merged in an official branch yet so we do not need to remember that you originally had a typo.

Yep I was planning to after it was ready to merge.

> indent the lines in your commit message

Do you mean in the hyphenated list?

File: org-msg.el [hunk 7]
```diff
@@ -155,8 +155,19 @@ It is used by function advice.")
...
+  "Alist of the available alternative exporters.
+Entries are in the form of `(tag . (mime-part . export-function))'.
+The export function exports the current `org-msg' edit buffer and
+returns the exported content as a string"
+  :type '(list sexp))
+
+(defcustom org-msg-default-alternatives '(html)
+  "List of alternative MIME formats to send.
+Available exporters are present in `org-msg-alternative-exporters'"
+  :type '(list symbol))
```
Comment (TimQuelch) [inline org-msg.el:170]: Is this an appropriate interface for the customise options?

**Comment by TimQuelch** [comment]
I've modified it to use the a configurable list that should be a little bit more extensible. Let me know if you have any thoughts

**Review by jeremy-compostella — COMMENTED** [review]
I posted a few comments.

File: org-msg.el (hunk 7 above)
Comment (jeremy-compostella) [reply org-msg.el:170]: Yes, note that we could rename `org-msg-alternative-exporters` into `org-msg-alternative`

Comment (TimQuelch) [reply org-msg.el:170]: I'm currently using `org-msg-alternatives` as the temp global variable that is used to store the content (for `org-msg-mml-into-multipart-related`

File: org-msg.el [hunk 8]
```diff
@@ -730,15 +732,35 @@ With the prefix argument ARG set, it calls
...
+
+(defun org-msg-get-message-attachments ()
+  "Get and validate the message attachments property."
+  (let ((attachments (org-msg-get-prop "attachments")))
+    (dolist (file attachments)
+      (unless (file-exists-p file)
+	(error "File '%s' does not exist" file)))
+    attachments))
+
+(defun org-msg-generate-alternatives (alternatives)
```
[the remark this answers is not in the document]
Comment (TimQuelch) [reply org-msg.el:762]: > It think you should be able to squash org-msg-get-message-alternatives and org-msg-get-available-alternatives into the single org-msg-generate-alternatives function

This is how I actually had it originally. I changed it because I thought that getting the list of available alternatives as well as extracting the current property value would be valuable functionality for future extension. e.g. I was planning on implementing a command to cycle through the available alternative types; this function would likely use both of these extracted functions. I think separating the functionality also will make `org-msg-generate-alternatives` easier to reason about.

> I would also call it org-msg-build-alternatives if you agree this is still an accurate name

Agreed

File: org-msg.el [hunk 9]
```diff
@@ -748,56 +770,43 @@ This function is a hook for `message-send-hook'."
...
-		    (mml-insert-tag 'part
-                                    'type (org-msg-file-mime-type file)
-				    'filename file
-                                    'disposition "attachment"))))
+                (dolist (alt org-msg-alternatives)
+                  (mml-insert-part (car alt))
+                  (insert (cdr alt))
+                  (forward-line))
+                (when (> (length alternatives) 1)
+                  (forward-line))
```
[the remark this answers is not in the document]
Comment (TimQuelch) [reply org-msg.el]: IMO yes. Semantically, including a multipart/alternative when there is only a single alternative is unnecessary and adds unneeded noise to the resulting MIME.

I don't believe sacrificing a better quality output for slightly simpler code is necessarily worth it.

In emacs lisp, is there a way to conditionally 'wrap' a bit of code? For example by default I want to,
```elisp
(dolist (alt org-msg-alternatives)
  (mml-insert-part (car alt))
  (insert (cdr alt))
  forward-line))
```
however if, the length of org-msg-alternatives is greater than 1 I want to wrap that whole statement in,
```
(mml-insert-multipart "alternative")
(<original code>)
(forward-line)
```
This is basically what I have now on two levels (if there are attachments, and if there are multiple alternatives), however the way I have it now results in evaluating the condition twice (before and after the 'inner' code). I'm sure there is a lispy way of doing this simpler but I can't think of one (at least not one that is more readable than the procedural way I have done it already)

**Comment by jeremy-compostella** [comment]
Do you mind squashing patches so that I can perform a review?

File: org-msg.el [hunk 10]
```diff
@@ -729,29 +774,48 @@ This function is a hook for `message-send-hook'."
...
-	  (dolist (file attachments)
-	    (unless (file-exists-p file)
-	      (error "File '%s' does not exist" file)))
+	(let ((alternatives (org-msg-get-message-alternatives))
+	      (attachments (org-msg-get-message-attachments)))
+          ;; Set temporary global variable with attachment list
 	  (setq org-msg-attachment attachments)
-	  (when org-msg-text-plain-alternative
-	    (setq org-msg-text-plain (org-msg-org-to-text-plain)))
+          ;; Generate contents of alternatives and save globally
```
Comment (jeremy-compostella) [inline org-msg.el:781]: indentation.

File: org-msg.el [hunk 11]
```diff
@@ -729,29 +774,48 @@ This function is a hook for `message-send-hook'."
...
+                  (insert (cdr alt))
+                  (forward-line))
+                (when (> (length alternatives) 1)
+                  (forward-line))
+                (dolist (file attachments)
+		  (mml-insert-tag 'part
+                                  'type (org-msg-file-mime-type file)
+				  'filename file
+                                  'disposition "attachment")))
+            (progn
```
Comment (jeremy-compostella) [inline org-msg.el:811]: `progn` is unnecessary is the else branch.

File: org-msg.el [hunk 12]
```diff
@@ -729,29 +774,48 @@ This function is a hook for `message-send-hook'."
...
-	  (mml-insert-part "text/html")
-	  (insert (propertize (org-msg-xml-to-str mail) 'mml t)))))))
+          ;; If mml has recursive html support (in later versions of emacs)
+          ;; we want to generate the structure of the MIME document here.
+          ;; If not we do this by manually editing the structure of the
+          ;; parsed MML tree in `org-msg-mml-into-multipart-related'. We
+          ;; also don't need to worry about this if we are only sending
+          ;; text/plain
+          (if (or (org-msg-mml-recursive-support)
+                  (not (member 'html alternatives)))
```
Comment (jeremy-compostella) [inline org-msg.el:794]: 1. member test with `equal`.  We already know we are manipulating symbol so `assq' is more appropriate.
2. Why do you exclude the `html` case here ?

File: org-msg.el [hunk 13]
```diff
@@ -854,8 +927,9 @@ automatically greet the right name, see `org-msg-greeting-fmt'."
 REPLY-TO is the file path of the original email export in HTML."
   (concat (format "#+OPTIONS: %s d:nil\n#+STARTUP: %s\n"
 		  (or org-msg-options "") (or org-msg-startup ""))
-	  (format ":PROPERTIES:\n:reply-to: %S\n:attachment: nil\n:END:\n"
-		  reply-to)))
+	  (format ":PROPERTIES:\n:reply-to: %S\n:attachment: nil\n:alternatives: %s\n:END:\n"
+		  reply-to
```
Comment (jeremy-compostella) [inline org-msg.el:931]: indentation. newline ?

File: org-msg.el [hunk 14]
```diff
@@ -729,29 +774,48 @@ This function is a hook for `message-send-hook'."
...
+                  (mml-insert-multipart "mixed"))
+                (when (> (length alternatives) 1)
+                  (mml-insert-multipart "alternative"))
+                (dolist (alt org-msg-alternatives)
+                  (mml-insert-part (car alt))
+                  (insert (cdr alt))
+                  (forward-line))
+                (when (> (length alternatives) 1)
+                  (forward-line))
+                (dolist (file attachments)
```
Comment (jeremy-compostella) [inline org-msg.el:806]: indentation ? Note that it could be just the gitbuh code renderer.

File: org-msg.el [hunk 15]
```diff
@@ -729,29 +774,48 @@ This function is a hook for `message-send-hook'."
...
+		  (mml-insert-tag 'part
```
Comment (jeremy-compostella) [inline org-msg.el:807]: Why is this on four lines now ?

File: org-msg.el (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeremy-compostella) [reply org-msg.el:760]: > This is how I actually had it originally. I changed it because I thought that getting the list of available alternatives as well as extracting the current property value would be valuable functionality for future extension. e.g. I was planning on implementing a command to cycle through the available alternative types; this function would likely use both of these extracted functions. I think separating the functionality also will make `org-msg-generate-alternatives` easier to reason about.

Unless you are already implementing it, I am usually against anticipating too much. Per my experience, anticipating just lead to useless and more complicated code. Very often what was anticipated isn't what is needed down the road. Plus I don't see how it necessarily makes the code cleaner here. I would rather see a code rework preceding or included with the new feature patch.

**Review by jeremy-compostella — COMMENTED** [review]
In addition to my comments, remember to update the documentation to reflect these changes.

File: org-msg.el (hunk 8 above)
[the remark this answers is not in the document]
Comment (TimQuelch) [reply org-msg.el:760]: Fair enough, I'll squash those functions

File: org-msg.el (hunk 12 above)
Comment (TimQuelch) [reply org-msg.el:783]: > member test with equal. We already know we are manipulating symbol so `assq' is more appropriate.
Oh nice I didn't know you could use `assq` as a member test for symbols. I thought it was mostly for alists

> Why do you exclude the html case here ?
As I understand it (which may be completely wrong...), `org-msg-mml-into-multipasrt-related` transforms the MIME structure of the HTML part into all of the required parts. However if there is no HTML part then this is not used. In this case we need to generate all the parts manually. The other case where we can generate the parts manually is when the `org-msg-mml-recursive-support` is there so that `mml` can expand the html part even if it is in a mixed/alternative multipart.

**Comment by TimQuelch** [comment]
I think many of the indentation errors are because there is a very large mix of tabs and spaces throughout the file. The style of indentation changes line to line.

File: org-msg.el (hunk 12 above)
Comment (jeremy-compostella) [reply org-msg.el]: Ooooh I see. It makes sense. I am looking forward to review the final squashed version. Don't forget to update the documentation (`README.org`).

File: org-msg.el (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeremy-compostella) [reply org-msg.el]: Unfortunately, I could not come up with a good solution either.  The thing is we are dealing with procedural (side-effects) functions.  As thus we are limited because the nice functional tricks do not apply in this case.

The only solution would be to save the point. Once the loop is over and if we have multiple alternatives we would do a  save-excursion, go back to the point, insert alternative and then call forward-line.  This is clearly going to be even uglier than what you currently have.

**Comment by TimQuelch** [comment]
I think it should be ready to merge.

I haven't fixed indentation because of the mix of tabs/spaces throughout the file. I think he best way to approach that would be a separate PR which fixes indentation for the whole file. Also including a .dir-locals which sets `indent-tabs-mode` to the desired value

File: org-msg.el [hunk 16]
```diff
@@ -155,8 +155,19 @@ It is used by function advice.")
...
 
-(defcustom org-msg-text-plain-alternative nil
-  "Include an ASCII export as a text/plain alternative.")
+(defcustom org-msg-alternative-exporters
+  '((text . ("text/plain" . org-msg-export-as-text))
+    (html . ("text/html" . org-msg-export-as-html)))
+  "Alist of the available alternative exporters.
+Entries are in the form of `(tag . (mime-part . export-function))'.
+The export function exports the current `org-msg' edit buffer and
+returns the exported content as a string"
```
Comment (jeremy-compostella) [inline org-msg.el:164]: add a point to finish the sentence in your comments.

**Review by jeremy-compostella — COMMENTED** [review]
Only one very minor comment but I would rather see it integrated than having to make a dedicated patch for it.

File: org-msg.el (hunk 16 above)
Comment (TimQuelch) [reply org-msg.el:164]: Oops that docstring is out of date too. Those functions accept a string now :o

**Comment by jeremy-compostella** [comment]
Thanks for your contribution.  I really appreciate that you took the time to take into my comments and improve your implementation.

**Comment by TimQuelch** [comment]
No worries, thanks for accepting the PR

**Comment by arvindsv** [comment]
FYI: I mentioned a potential surprise due to this change, which might be useful to others looking at this PR at a later time: https://github.com/hlissner/doom-emacs/pull/4295#issuecomment-737993746
